packages/webpack-loader/src/compiled-loader.tsx

```diff
@@ -0,0 +1,121 @@
+import path from 'path';
+import { transformFromAstAsync, parseAsync } from '@babel/core';
+import { getOptions } from 'loader-utils';
+import { toBoolean, createError } from '@compiled/utils';
+import type { CompiledLoaderOptions, LoaderThis } from './types';
+import { pluginName } from './extract-plugin';
+
+/**
+ * Returns user configuration.
+ *
+ * @param loader
+ * @returns
+ */
+function getLoaderOptions(context: LoaderThis<CompiledLoaderOptions>) {
+  const options: CompiledLoaderOptions =
+    typeof context.getOptions === 'undefined'
+      ? // Webpack v4 flow
+        getOptions(context)
+      : // Webpack v5 flow
+        context.getOptions({
+          type: 'object',
+          properties: {
+            importReact: {
+              type: 'boolean',
+            },
+            nonce: {
+              type: 'string',
+            },
+            extract: {
+              type: 'boolean',
+            },
+          },
+        });
+
+  return options;
+}
+
+/**
+ * Compiled webpack loader.
+ *
+ * @param this
+ * @param code
+ */
+export default async function compiledLoader(
+  this: LoaderThis<CompiledLoaderOptions>,
+  code: string
+): Promise<void> {
+  const callback = this.async();
+
+  // Bail early if Compiled isn't in the module.
+  if (code.indexOf('@compiled/react') === -1) {
+    return callback(null, code);
+  }
+
+  try {
+    const includedFiles: string[] = [];
+    const foundCSSRules: string[] = [];
+    const options = getLoaderOptions(this);
+
+    // Transform to an AST using the local babel config.
+    const ast = await parseAsync(code, {
+      filename: this.resourcePath,
+      caller: { name: 'compiled' },
+    });
+
+    // Transform using the Compiled Babel Plugin - we deliberately turn off using the local config.
+    const result = await transformFromAstAsync(ast!, code, {
+      babelrc: false,
+      configFile: false,
+      sourceMaps: true,
+      filename: this.resourcePath,
+      plugins: [
+        options.extract && [
+          '@compiled/babel-plugin-strip-runtime',
+          { onFoundStyleRules: (rules: string[]) => foundCSSRules.push(...rules) },
+        ],
+        [
+          '@compiled/babel-plugin',
+          { ...options, onIncludedFiles: (files: string[]) => includedFiles.push(...files) },
+        ],
+      ].filter(toBoolean),
+    });
+
+    includedFiles.forEach((file) => {
+      this.addDependency(path.normalize(file));
+    });
+
+    let output: string = result?.code || '';
+
+    if (options.extract && foundCSSRules.length) {
+      foundCSSRules.forEach((rule) => {
+        // Each found atomic rule will create a new import that uses `@compiled/webpack-loader/css-loader`.
```

Comment:
So are we creating import for each atomic rule? Any disadvantage of this if we have many atomic rules? My understanding is limited here, so just clarifying 😅